### DISCUSSION TITLE
Web3 Wallet Auth

### DISCUSSION THREAD

**Discussion by Rahat-ch** [supabase/supabase#4713] [Feature Requests]
# Feature request

## Is your feature request related to a problem? Please describe.

Log in with Web3 wallets - starting with EVM(Ethereum) based wallets. 
This allows traditionally web2 applications to slowly start tapping into the web3 ecosystem by first allowing users to bring their web3 identity with them into their application instead of creating a new user/account or password or having to choose to have it managed by a third party auth service. 

## Describe the solution you'd like
Similar to other auth methods we should allow users we should have easy to use functionality that allows developers to implement web3 login modals into their apps from supabase to quickly allow their users to sign a transaction and be authenticated by supabase using an evm wallet. 

We would need frontend to begin the request - backend to generate a Nonce(number only used once) and associate it to the user account if it already exists or create a new user account based on the address if it doesn't exist and then generate the Nonce. The server would send the Nonce to the frontend - the frontend will sign via their wallet and a crypto graphical signature would be generated using the address and the nonce. The address would be sent back to the server and be parsed to verify the correct address and nonce are associated to the signature and then continue with regular auth process.

## Describe alternatives you've considered

We can also support Solana wallets and other chains as well but with Ethereum being the most popular I think first iterations should be focused there.

## Additional context

Looking through current auth implementations to see best way of implementing this would love to lead this effort.

**Comment by HarryET** [comment]
Have a look at:
- https://github.com/supabase/gotrue/pull/282
- https://twitter.com/TheHarryET/status/1461725915071881217

**Reply by tansanDOTeth** [reply]
I took a look at this, and I don't think you ever need to store the nonce. That defeats the ephemeral nature of a nonce. Could it be simplified so it can pass a review?

**Reply by HarryET** [reply]
This PR is being closed by the Supabase team, I will flag with them so I can have it closed because there are no plans for this to be merged

**Comment by devstein** [comment]
[Picket](https://picketapi.com/) is a good fit for this use-case. It supports wallet-based authentication on both Ethereum and Solana

**Comment by nattstack** [comment]
Seems like native web3 wallet auth won't come in Supabase for a while (see https://github.com/supabase/gotrue/pull/282). In the meanwhile, a solution would be using a third party library to sign in with wallet then store the wallet address as metadata or new table with foreign key with `auth.users.id` and the mental modal of wallet address would act as a username.

Good third party libraries for wallet sign in
- https://github.com/rainbow-me/rainbowkit
- https://github.com/tmm/wagmi

**Reply by HarryET** [reply]
This is the PR your refering too, https://github.com/supabase/gotrue/pull/282, and a good comment from the supabse team is on it too.

**Comment by tansanDOTeth** [comment]
Since it's just a simple signing and issuance of a token, is it possible to do that with the existing infrastructure in Supabase? 

What I would like is something like:
* Send signed nonce
* Creates a user record if it doesn't exist already with the wallet id
* Sends back a token

I'm trying to save myself from having to create a whole backend.

**Reply by afeezaziz** [reply]
I would concur with this idea.

**Comment by bakeiro** [comment]
Hi!
I can see in the docs that sign in with ethereum is "coming soon", does anyone knows anything about this? I don't see any feedback from the supabase team about this (none release date, log, or anything)

**Reply by cemalkilic** [reply]
Hi @bakeiro, the changes on the auth server side is done (https://github.com/supabase/auth/pull/2069). We need to update the dashboard UI to let users configure ethereum, plus update docs.
Probably going to be fully released in a few weeks, sorry for the delay!

**Reply by bakeiro** [reply]
oh wow, that sounds incredible, thanks!!!! (and extremely convenient since I needed it lol)

**Reply by nathanvolk-dev** [reply]
@cemalkilic sorry to sound dumb here, but when you say configure Ethereum, does this include Metamask login? Via the chrome extension ?

**Reply by cemalkilic** [reply]
@nathanvolk-dev by configuring, I meant to enable Ethereum login in the [Supabase dashboard](https://supabase.com/dashboard/project/_/auth/providers). And yes, the _end users_ will be able to use Metamask login.
